ang/civiawards-workflow/services/applicant-management-workflow.service.js

```diff
@@ -0,0 +1,156 @@
+
+          return civicaseCrmApi('AwardDetail', 'get', filters)
+            .then(function (awardsData) {
+              return _.map(awardsData.values, function (award) {
+                return award.case_type_id;
+              });
+            });
+        });
+    }
+
+    /**
+     * Returns workflows list for applicant management
+     *
+     * @param {Array} scope scope object of the workflows list controller
+     * @returns {Array} api call parameters
+     */
+    function getWorkflowsList (scope) {
+      return applyFilter(scope)
+        .then(function (caseTypeIds) {
+          var params = {
+            sequential: 1,
+            is_active: scope.selectedFilters.is_active,
+            case_type_category: scope.caseTypeCategory,
+            options: { limit: 0 },
+            'api.AwardDetail.get': {
+              case_type_id: '$value.id'
+            }
+          };
+
+          if (caseTypeIds) {
+            if (caseTypeIds.length === 0) {
+              return $q.resolve([]);
+            }
+            params.id = { IN: caseTypeIds };
+          }
+
+          return civicaseCrmApi('CaseType', 'get', params).then(function (data) {
+            return getFormattedAwardDetailsData(data.values);
+          });
+        });
+    }
+
+    /**
+     * @param {object} workflow workflow object
+     * @returns {Promise} promise
+     */
+    function createDuplicate (workflow) {
+      return civicaseCrmApi('CaseType', 'create', _.extend({}, workflow, { id: null }))
+        .then(function (data) {
+          var awardsDetailsData = _.clone(workflow.awardDetails);
+
+          awardsDetailsData.id = null;
+          awardsDetailsData.case_type_id = data.values[0].id;
+
+          return civicaseCrmApi([
+            ['AwardDetail', 'create', awardsDetailsData]
+          ]);
+        });
+    }
+
+    /**
+     * @param {object[]} workflows list of workflows
+     * @returns {object[]} formatted list of workflows
+     */
+    function getFormattedAwardDetailsData (workflows) {
+      var managerContacts = [];
+
+      workflows = _.map(workflows, function (workflow) {
+        workflow.awardDetails = workflow['api.AwardDetail.get'].values[0];
+        delete workflow['api.AwardDetail.get'];
+
+        managerContacts = managerContacts.concat(workflow.awardDetails.award_manager);
+
+        workflow.awardDetailsFormatted = {
+          subtypeLabel: awardSubtypes[workflow.awardDetails.award_subtype].label,
+          managers: []
+        };
+
+        return workflow;
+      });
+
+      return ContactsCache.add(managerContacts)
+        .then(function () {
+          return _.map(workflows, function (workflow) {
+            _.each(workflow.awardDetails.award_manager, function (managerID) {
+              workflow.awardDetailsFormatted.managers.push(
+                ContactsCache.getCachedContact(managerID).display_name
+              );
+            });
+
+            return workflow;
+          });
+        });
+    }
```

Comment:
This function is doing too much so it's a bit confusing to follow it. Specially because we are adding the managers here, and the `ContactsCache.add` function does not help. First of all, it should be called `fetch`, not `add`, and second, it should be called elsewhere.

I think this function and the `getWorkflowsList` function should be refactored as follow:

```js
function getWorkflowsList (filters) {
  return fetchFilteredAwardTypeIds(filters)
    .then(function (awardTypeIds) {
      return fetchWorkflows(_.assign({}, filters, {
        awardTypeIds: awardTypeIds
      });
    })
    .then(function (workflows) {
      var awardManagers = getAwardManagersForWorkflows(workflows);

      // we fetch all the contacts in one go instead of making multiple calls:
      return ContactsCache.add(awardManagers)
        .then(function () {
          return _.map(awardTypes, getFormattedAwardDetailsData);
        });
    });
  
}

function getAwardManagersForWorkflows (workflows) {
  return _.chain(workflows)
    .map(function (workflow) {
      return workflow['api.AwardDetail.get'].values[0]. award_manager;
    })
    .flatten()
    .value();
}

function getFormattedAwardDetailsData (workflow) {
  var awardDetails = workflow['api.AwardDetail.get'].values[0];
  var managers = _.map(awardDetails.award_manager, function (managerId) {
    return ContactsCache.getCachedContact(managerId).display_name;
  });

  return {
    // any fields we are interested in
    awardDetails: awardDetails,
    formattedAwardDetails: {
      subtypeLabel: awardSubtypes[awardDetails.award_subtype].label,
      managers: managers
    },
  };
}
```

Reply:
I have refactored this whole thing, and now it looks more organised. Though I have not followed this code snippet completely. So lets check it once more, to see if further refactoring is necessary.